From 44de9e746525c678c8c9a32dd6b79684dd97582b Mon Sep 17 00:00:00 2001
From: Aliaksei Kandratsenka <alkondratenko@gmail.com>
Date: Fri, 16 Oct 2026 11:39:05 -0400
Subject: [PATCH 3/4] elf: Move dlopen/dlclose _dl_find_object updates out of
 dl_load_tls_lock

Move the fallible _dl_find_object_update call, together with
activate_nodelete, update_scopes, the reloc_complete probe and the
libc early init call, from dl_open_worker_begin to dl_open_worker,
after dl_load_tls_lock has been released.  The _dl_find_object_update
call now comes before the demarcation point, since it can fail.

The RT_CONSISTENT debugger/audit checkpoint is moved into
maybe_report_r_debug_consistent, which is called on the success path
before initializers run, and from _dl_open for the error path, and
ensures it is signalled exactly once.  The r_debug pointer used for the
reloc_start/reloc_complete probes is now kept in struct dl_open_args.

In dlclose, call _dl_find_object_dlclose for all unloaded objects in a
separate loop before the TLS and link map teardown.

Signed-off-by: Aliaksei Kandratsenka <alkondratenko@gmail.com>
---
 elf/dl-close.c       |  13 +++-
 elf/dl-find_object.h |   4 +-
 elf/dl-open.c        | 148 +++++++++++++++++++++++++------------------
 3 files changed, 98 insertions(+), 67 deletions(-)

diff --git a/elf/dl-close.c b/elf/dl-close.c
index 8b6e654791..39e1cf89ff 100644
--- a/elf/dl-close.c
+++ b/elf/dl-close.c
@@ -454,60 +454,70 @@ _dl_close_worker (struct link_map *map, bool force)
 
       if (cnt + unload_global == ns_msl->r_nlist)
 	/* Speed up removing most recently added objects.  */
 	j = cnt;
       else
 	for (i = 0; i < cnt; i++)
 	  if (ns_msl->r_list[i]->l_removed == 0)
 	    {
 	      if (i != j)
 		ns_msl->r_list[j] = ns_msl->r_list[i];
 	      j++;
 	    }
       ns_msl->r_nlist = j;
     }
 
   if (!RTLD_SINGLE_THREAD_P
       && (unload_global
 	  || scope_mem_left
 	  || (GL(dl_scope_free_list) != NULL
 	      && GL(dl_scope_free_list)->count)))
     {
       THREAD_GSCOPE_WAIT ();
 
       /* Now we can free any queued old scopes.  */
       struct dl_scope_free_list *fsl = GL(dl_scope_free_list);
       if (fsl != NULL)
 	while (fsl->count > 0)
 	  free (fsl->list[--fsl->count]);
     }
 
+  for (unsigned int i = first_loaded; i < nloaded; ++i)
+    {
+      struct link_map *imap = maps[i];
+      if (imap->l_map_used)
+	continue;
+
+      /* Update the data used by _dl_find_object.  */
+      _dl_find_object_dlclose (imap);
+    }
+
   size_t tls_free_start;
   size_t tls_free_end;
   tls_free_start = tls_free_end = NO_TLS_OFFSET;
 
   /* Protects global and module specific TLS state.  */
   __rtld_lock_lock_recursive (GL(dl_load_tls_lock));
 
   /* We modify the list of loaded objects.  */
   __rtld_lock_lock_recursive (GL(dl_load_write_lock));
 
   /* Check each element of the search list to see if all references to
      it are gone.  */
   for (unsigned int i = first_loaded; i < nloaded; ++i)
     {
       struct link_map *imap = maps[i];
       if (!imap->l_map_used)
 	{
 	  assert (imap->l_type == lt_loaded);
 
 	  /* That was the last reference, and this was a dlopen-loaded
 	     object.  We can unmap it.  */
 
 	  /* Remove the object from the dtv slotinfo array if it uses TLS.  */
 	  if (__glibc_unlikely (imap->l_tls_blocksize > 0))
 	    {
 	      any_tls = true;
 
 	      if (GL(dl_tls_dtv_slotinfo_list) != NULL
 		  && ! remove_slotinfo (imap->l_tls_modid,
 					GL(dl_tls_dtv_slotinfo_list), 0,
@@ -633,63 +643,60 @@ _dl_close_worker (struct link_map *map, bool force)
 	     the `munmap' call does the rest.  */
 	  DL_UNMAP (imap);
 
 	  /* Finally, unlink the data structure and free it.  */
 #if DL_NNS == 1
 	  /* The assert in the (imap->l_prev == NULL) case gives
 	     the compiler license to warn that NS points outside
 	     the dl_ns array bounds in that case (as nsid != LM_ID_BASE
 	     is tantamount to nsid >= DL_NNS).  That should be impossible
 	     in this configuration, so just assert about it instead.  */
 	  assert (nsid == LM_ID_BASE);
 	  assert (imap->l_prev != NULL);
 #else
 	  if (imap->l_prev == NULL)
 	    {
 	      assert (nsid != LM_ID_BASE);
 	      ns->_ns_loaded = imap->l_next;
 
 	      /* Update the pointer to the head of the list
 		 we leave for debuggers to examine.  */
 	      r->r_map = (void *) ns->_ns_loaded;
 	    }
 	  else
 #endif
 	    imap->l_prev->l_next = imap->l_next;
 
 	  --ns->_ns_nloaded;
 	  if (imap->l_next != NULL)
 	    imap->l_next->l_prev = imap->l_prev;
 
-	  /* Update the data used by _dl_find_object.  */
-	  _dl_find_object_dlclose (imap);
-
 	  free (imap->l_versions);
 	  if (imap->l_origin != (char *) -1)
 	    free ((char *) imap->l_origin);
 
 	  free (imap->l_reldeps);
 
 	  /* Print debugging message.  */
 	  if (__glibc_unlikely (GLRO(dl_debug_mask) & DL_DEBUG_FILES))
 	    _dl_debug_printf ("\nfile=%s [%lu];  destroying link map\n",
 			      imap->l_name, imap->l_ns);
 
 	  /* This name always is allocated.  */
 	  free (imap->l_name);
 	  /* Remove the list with all the names of the shared object.  */
 
 	  struct libname_list *lnp = imap->l_libname;
 	  do
 	    {
 	      struct libname_list *this = lnp;
 	      lnp = lnp->next;
 	      if (!this->dont_free)
 		free (this);
 	    }
 	  while (lnp != NULL);
 
 	  /* Remove the searchlists.  */
 	  free (imap->l_initfini);
 
 	  /* Remove the scope array if we allocated it.  */
 	  if (imap->l_scope != imap->l_scope_mem)
diff --git a/elf/dl-find_object.h b/elf/dl-find_object.h
index 9314c84432..70027c3e9e 100644
--- a/elf/dl-find_object.h
+++ b/elf/dl-find_object.h
@@ -129,42 +129,42 @@ _dl_find_object_from_map (struct link_map *l,
 #if DLFO_STRUCT_HAS_EH_COUNT
           atomic_store_relaxed (&result->eh_count, ph->p_memsz / 8);
 #endif
           read_seg |= 1;
           break;
 
         case PT_GNU_SFRAME:
           atomic_store_relaxed (&result->sframe,
                                 (void *) (ph->p_vaddr + l->l_addr));
           read_seg |= 2;
 	  [[fallthrough]];
         default:
           break;
         }
       if (read_seg == 3)
         goto done;
    }
 
  done:
   l->l_find_object_processed = 1;
 }
 
 /* Called by the dynamic linker to set up the data structures for the
    initially loaded objects.  This creates a few persistent
    allocations, so it should be called with the minimal malloc.  */
 void _dl_find_object_init (void) attribute_hidden;
 
 /* Called by dlopen/dlmopen to add new objects to the DWARF EH frame
    data structures.  NEW_MAP is the dlopen'ed link map.  Link maps on
    the l_next list are added if l_object_processed is 0.  Needs to
-   be protected by loader write lock.  Returns true on success, false
+   be protected by loader load lock.  Returns true on success, false
    on malloc failure.  */
 bool _dl_find_object_update (struct link_map *new_map) attribute_hidden;
 
 /* Called by dlclose to remove the link map from the DWARF EH frame
-   data structures.  Needs to be protected by loader write lock.  */
+   data structures.  Needs to be protected by loader load lock.  */
 void _dl_find_object_dlclose (struct link_map *l) attribute_hidden;
 
 /* Called from __libc_freeres to deallocate malloc'ed memory.  */
 void _dl_find_object_freeres (void) attribute_hidden;
 
 #endif /* _DL_FIND_OBJECT_H */
diff --git a/elf/dl-open.c b/elf/dl-open.c
index 848b1afd2c..21d589b12e 100644
--- a/elf/dl-open.c
+++ b/elf/dl-open.c
@@ -42,60 +42,69 @@
 #include <dl-prop.h>
 
 
 /* We must be careful not to leave us in an inconsistent state.  Thus we
    catch any error and re-raise it after cleaning up.  */
 
 struct dl_open_args
 {
   const char *file;
   int mode;
   struct link_map *caller_map; /* Derived from the caller address.  */
   struct link_map *map;
   /* Namespace ID.  */
   Lmid_t nsid;
 
   /* Original value of _ns_global_scope_pending_adds.  Set by
      dl_open_worker.  Only valid if nsid is a real namespace
      (non-negative).  */
   unsigned int original_global_scope_pending_adds;
 
   /* Set to true by dl_open_worker if libc.so was already loaded into
      the namespace at the time dl_open_worker was called.  This is
      used to determine whether libc.so early initialization has
      already been done before, and whether to roll back the cached
      libc_map value in the namespace in case of a dlopen failure.  */
   bool libc_already_loaded;
 
   /* Set to true if the end of dl_open_worker_begin was reached.  */
   bool worker_continue;
 
+  /* Set once the RT_CONSISTENT debugger/audit checkpoint has been
+     signalled, so that it happens exactly once on both the success
+     and the error path.  */
+  bool reported_r_debug_consistent;
+
+  /* Set to r_debug pointer after reloc_start probe is signaled. Used
+     to signal reloc_complete probe.  */
+  struct r_debug *relocation_probe_r_debug;
+
   /* Original parameters to the program and the current environment.  */
   int argc;
   char **argv;
   char **env;
 };
 
 /* Called in case the global scope cannot be extended.  */
 static void __attribute__ ((noreturn))
 add_to_global_resize_failure (struct link_map *new)
 {
   _dl_signal_error (ENOMEM, new->l_libname->name, NULL,
 		    N_ ("cannot extend global scope"));
 }
 
 /* Grow the global scope array for the namespace, so that all the new
    global objects can be added later in add_to_global_update, without
    risk of memory allocation failure.  add_to_global_resize raises
    exceptions for memory allocation errors.  */
 static void
 add_to_global_resize (struct link_map *new)
 {
   struct link_namespaces *ns = &GL (dl_ns)[new->l_ns];
 
   /* Count the objects we have to put in the global scope.  */
   unsigned int to_add = 0;
   for (unsigned int cnt = 0; cnt < new->l_searchlist.r_nlist; ++cnt)
     if (new->l_searchlist.r_list[cnt]->l_global == 0)
       ++to_add;
 
   /* The symbols of the new objects and its dependencies are to be
@@ -368,75 +377,73 @@ resize_tls_slotinfo (struct link_map *new)
 }
 
 /* Mark the objects as NODELETE if required.  This is delayed until
    after dlopen failure is not possible, so that _dl_close can clean
    up objects if necessary.  */
 static void
 activate_nodelete (struct link_map *new)
 {
   /* It is necessary to traverse the entire namespace.  References to
      objects in the global scope and unique symbol bindings can force
      NODELETE status for objects outside the local scope.  */
   for (struct link_map *l = GL (dl_ns)[new->l_ns]._ns_loaded; l != NULL;
        l = l->l_next)
     if (l->l_nodelete_pending)
       {
 	if (__glibc_unlikely (GLRO (dl_debug_mask) & DL_DEBUG_FILES))
 	  _dl_debug_printf ("activating NODELETE for %s [%lu]\n",
 			    l->l_name, l->l_ns);
 
 	/* The flag can already be true at this point, e.g. a signal
 	   handler may have triggered lazy binding and set NODELETE
 	   status immediately.  */
 	l->l_nodelete_active = true;
 
 	/* This is just a debugging aid, to indicate that
 	   activate_nodelete has run for this map.  */
 	l->l_nodelete_pending = false;
       }
 }
 
-/* Relocate the object L.  *RELOCATION_IN_PROGRESS controls whether
-   the debugger is notified of the start of relocation processing.  */
+/* Relocate the object L. */
 static void
 _dl_open_relocate_one_object (struct dl_open_args *args, struct r_debug *r,
-			      struct link_map *l, int reloc_mode,
-			      bool *relocation_in_progress)
+			      struct link_map *l, int reloc_mode)
 {
   if (l->l_real->l_relocated)
     return;
 
-  if (!*relocation_in_progress)
+  if (!args->relocation_probe_r_debug)
     {
       /* Notify the debugger that relocations are about to happen.  */
       LIBC_PROBE (reloc_start, 2, args->nsid, r);
-      *relocation_in_progress = true;
+      args->relocation_probe_r_debug = r;
     }
 
 #ifdef SHARED
   if (__glibc_unlikely (GLRO(dl_profile) != NULL))
     {
       /* If this here is the shared object which we want to profile
 	 make sure the profile is started.  We can find out whether
 	 this is necessary or not by observing the `_dl_profile_map'
 	 variable.  If it was NULL but is not NULL afterwards we must
 	 start the profiling.  */
       struct link_map *old_profile_map = GL(dl_profile_map);
 
       _dl_relocate_object (l, l->l_scope, reloc_mode | RTLD_LAZY, 1);
 
       if (old_profile_map == NULL && GL(dl_profile_map) != NULL)
 	{
 	  /* We must prepare the profiling.  */
 	  _dl_start_profile ();
 
 	  /* Prevent unloading the object.  */
 	  GL(dl_profile_map)->l_nodelete_active = true;
 	}
     }
   else
 #endif
     _dl_relocate_object (l, l->l_scope, reloc_mode, 0);
 }
 
 static void
 call_dl_init (void *closure)
@@ -582,205 +589,217 @@ dl_open_worker_begin (void *a)
 
   /* Print scope information.  */
   if (__glibc_unlikely (GLRO(dl_debug_mask) & DL_DEBUG_SCOPES))
     _dl_show_scope (new, 0);
 
   /* Only do lazy relocation if `LD_BIND_NOW' is not set.  */
   int reloc_mode = mode & __RTLD_AUDIT;
   if (GLRO(dl_lazy))
     reloc_mode |= mode & RTLD_LAZY;
 
   /* Objects must be sorted by dependency for the relocation process.
      This allows IFUNC relocations to work and it also means copy
      relocation of dependencies are if necessary overwritten.
      __dl_map_object_deps has already sorted l_initfini for us.  */
   unsigned int first = UINT_MAX;
   unsigned int last = 0;
   unsigned int j = 0;
   struct link_map *l = new->l_initfini[0];
   do
     {
       if (! l->l_real->l_relocated)
 	{
 	  if (first == UINT_MAX)
 	    first = j;
 	  last = j + 1;
 	}
       l = new->l_initfini[++j];
     }
   while (l != NULL);
 
-  bool relocation_in_progress = false;
-
   /* This only performs the memory allocations.  The actual update of
      the scopes happens below, after failure is impossible.  */
   resize_scopes (new);
 
   /* Increase the size of the GL (dl_tls_dtv_slotinfo_list) data
      structure.  */
   bool any_tls = resize_tls_slotinfo (new);
 
   /* Perform the necessary allocations for adding new global objects
      to the global scope below.  */
   if (mode & RTLD_GLOBAL)
     add_to_global_resize (new);
 
   /* Register the new modules in the DTV slotinfo and bump the TLS
      generation counter *before* relocation, so an IFUNC resolver firing
      during the relocation loop below can reach its DSO's __thread storage
      via __tls_get_addr / TLSDESC.  Without this, the new module is not yet
      in GL(dl_tls_dtv_slotinfo_list), so the resolver's dynamic-TLS lookup
      fails to find it and faults.  The static-TLS image itself is copied
      lazily on first access, and if relocation later fails, the subsequent
      _dl_close_worker cleans up these slotinfo entries via remove_slotinfo.  */
   if (any_tls)
     {
       for (unsigned int i = 0; i < new->l_searchlist.r_nlist; ++i)
 	_dl_add_to_slotinfo (new->l_searchlist.r_list[i], true);
 
       size_t newgen = GL(dl_tls_generation) + 1;
       if (__glibc_unlikely (newgen == 0))
 	_dl_fatal_printf (N_("\
 TLS generation counter wrapped!  Please report this."));
       /* Can be read concurrently.  */
       atomic_store_release (&GL(dl_tls_generation), newgen);
     }
 
   /* Perform relocation.  This can trigger lazy binding in IFUNC
      resolvers.  For NODELETE mappings, these dependencies are not
      recorded because the flag has not been applied to the newly
      loaded objects.  This means that upon dlopen failure, these
      NODELETE objects can be unloaded despite existing references to
      them.  However, such relocation dependencies in IFUNC resolvers
      are undefined anyway, so this is not a problem.  */
 
   /* Ensure that libc is relocated first.  This helps with the
      execution of IFUNC resolvers in libc, and matters only to newly
      created dlmopen namespaces.  Do not do this for static dlopen
      because libc has relocations against ld.so, which may not have
      been relocated at this point.  */
   struct r_debug *r = _dl_debug_update (args->nsid);
 #ifdef SHARED
   if (GL(dl_ns)[args->nsid].libc_map != NULL)
     _dl_open_relocate_one_object (args, r, GL(dl_ns)[args->nsid].libc_map,
-				  reloc_mode, &relocation_in_progress);
+				  reloc_mode);
 #endif
 
   for (unsigned int i = last; i-- > first; )
-    _dl_open_relocate_one_object (args, r, new->l_initfini[i], reloc_mode,
-				  &relocation_in_progress);
-
-  /* Demarcation point: After this, no recoverable errors are allowed.
-     All memory allocations for new objects must have happened
-     before.  */
+    _dl_open_relocate_one_object (args, r, new->l_initfini[i], reloc_mode);
 
-  /* Finalize the NODELETE status first.  This comes before
-     update_scopes, so that lazy binding will not see pending NODELETE
-     state for newly loaded objects.  There is a compiler barrier in
-     update_scopes which ensures that the changes from
-     activate_nodelete are visible before new objects show up in the
-     local scope.  */
-  activate_nodelete (new);
-
-  /* Second stage after resize_scopes: Actually perform the scope
-     update.  After this, dlsym and lazy binding can bind to new
-     objects.  */
-  update_scopes (new);
-
-  if (!_dl_find_object_update (new))
-    _dl_signal_error (ENOMEM, new->l_libname->name, NULL,
-		      N_ ("cannot allocate address lookup data"));
+  args->worker_continue = true;
+}
 
-  /* Notify the debugger all new objects have been relocated.  */
-  if (relocation_in_progress)
-    LIBC_PROBE (reloc_complete, 3, args->nsid, r, new);
+static void
+maybe_report_r_debug_consistent (struct dl_open_args *args)
+{
+  if (args->reported_r_debug_consistent)
+    return;
 
-  /* If libc.so was not there before, attempt to call its early
-     initialization routine.  Indicate to the initialization routine
-     whether the libc being initialized is the one in the base
-     namespace.  */
-  if (!args->libc_already_loaded)
+  /* Auditing checkpoint and debugger signalling.  Do this even on
+     error, so that dlopen exits with consistent state.  */
+  if (args->nsid >= 0 || args->map != NULL)
     {
-      /* dlopen cannot be used to load an initial libc by design.  */
-      struct link_map *libc_map = GL(dl_ns)[args->nsid].libc_map;
-      _dl_call_libc_early_init (libc_map, false);
+      Lmid_t nsid = args->map != NULL ? args->map->l_ns : args->nsid;
+      struct r_debug *r = _dl_debug_update (nsid);
+#ifdef SHARED
+      bool was_not_consistent  = r->r_state != RT_CONSISTENT;
+#endif
+      _dl_debug_change_state (r, RT_CONSISTENT);
+      LIBC_PROBE (map_complete, 3, nsid, r, args->map);
+
+#ifdef SHARED
+      if (was_not_consistent)
+	/* Avoid redundant/recursive signalling.  */
+	_dl_audit_activity_nsid (nsid, LA_ACT_CONSISTENT);
+#endif
     }
 
-  args->worker_continue = true;
+  args->reported_r_debug_consistent = true;
 }
 
 static void
 dl_open_worker (void *a)
 {
   struct dl_open_args *args = a;
 
   args->worker_continue = false;
 
   {
     /* Protects global and module specific TLS state.  */
     __rtld_lock_lock_recursive (GL(dl_load_tls_lock));
 
     struct dl_exception ex;
     int err = _dl_catch_exception (&ex, dl_open_worker_begin, args);
 
     __rtld_lock_unlock_recursive (GL(dl_load_tls_lock));
 
-    /* Auditing checkpoint and debugger signalling.  Do this even on
-       error, so that dlopen exists with consistent state.  */
-    if (args->nsid >= 0 || args->map != NULL)
-      {
-	Lmid_t nsid = args->map != NULL ? args->map->l_ns : args->nsid;
-	struct r_debug *r = _dl_debug_update (nsid);
-#ifdef SHARED
-	bool was_not_consistent  = r->r_state != RT_CONSISTENT;
-#endif
-	_dl_debug_change_state (r, RT_CONSISTENT);
-	LIBC_PROBE (map_complete, 3, nsid, r, args->map);
-
-#ifdef SHARED
-	if (was_not_consistent)
-	  /* Avoid redundant/recursive signalling.  */
-	  _dl_audit_activity_nsid (nsid, LA_ACT_CONSISTENT);
-#endif
-      }
-
     if (__glibc_unlikely (ex.errstring != NULL))
       /* Reraise the error.  */
       _dl_signal_exception (err, &ex, NULL);
   }
 
   if (!args->worker_continue)
     return;
 
   int mode = args->mode;
   struct link_map *new = args->map;
 
+  /* Relocation is complete and none of the code below needs to keep
+     dl_load_tls_lock held.  */
+
+  if (!_dl_find_object_update (new))
+    _dl_signal_error (ENOMEM, new->l_libname->name, NULL,
+		      N_ ("cannot allocate address lookup data"));
+
+  /* Demarcation point: After this, no recoverable errors are allowed.
+     All memory allocations for new objects must have happened
+     before.  */
+
+  /* Finalize the NODELETE status first.  This comes before
+     update_scopes, so that lazy binding will not see pending NODELETE
+     state for newly loaded objects.  There is a compiler barrier in
+     update_scopes which ensures that the changes from
+     activate_nodelete are visible before new objects show up in the
+     local scope.  */
+  activate_nodelete (new);
+
+  /* Second stage after resize_scopes: Actually perform the scope
+     update.  After this, dlsym and lazy binding can bind to new
+     objects.  */
+  update_scopes (new);
+
+  /* Notify the debugger all new objects have been relocated.  */
+  if (args->relocation_probe_r_debug)
+    LIBC_PROBE (reloc_complete, 3, args->nsid,
+		args->relocation_probe_r_debug, new);
+
+  /* If libc.so was not there before, attempt to call its early
+     initialization routine.  Indicate to the initialization routine
+     whether the libc being initialized is the one in the base
+     namespace.  */
+  if (!args->libc_already_loaded)
+    {
+      /* dlopen cannot be used to load an initial libc by design.  */
+      struct link_map *libc_map = GL(dl_ns)[args->nsid].libc_map;
+      _dl_call_libc_early_init (libc_map, false);
+    }
+
+  maybe_report_r_debug_consistent (args);
+
   /* Run the initializer functions of new objects.  Temporarily
      disable the exception handler, so that lazy binding failures are
      fatal.  */
   _dl_catch_exception (NULL, call_dl_init, args);
 
   /* Now we can make the new map available in the global scope.  */
   if (mode & RTLD_GLOBAL)
     add_to_global_update (new);
 
   /* Let the user know about the opencount.  */
   if (__glibc_unlikely (GLRO(dl_debug_mask) & DL_DEBUG_FILES))
     _dl_debug_printf ("opening file=%s [%lu]; direct_opencount=%u\n\n",
 		      new->l_name, new->l_ns, new->l_direct_opencount);
 }
 
 void *
 _dl_open (const char *file, int mode, const void *caller_dlopen, Lmid_t nsid,
 	  int argc, char *argv[], char *env[])
 {
   if ((mode & RTLD_BINDING_MASK) == 0)
     /* One of the flags must be set.  */
     _dl_signal_error (EINVAL, file, NULL, N_("invalid mode for dlopen()"));
 
   /* Make sure we are alone.  */
   __rtld_lock_lock_recursive (GL(dl_load_lock));
 
   if (__glibc_unlikely (nsid == LM_ID_NEWLM))
     {
       /* Find a new namespace.  */
       for (nsid = 1; DL_NNS > 1 && nsid < GL(dl_nns); ++nsid)
@@ -798,95 +817,100 @@ no more namespaces available for dlmopen()"));
       else if (nsid == GL(dl_nns))
 	{
 	  __rtld_lock_initialize (GL(dl_ns)[nsid]._ns_unique_sym_table.lock);
 	  ++GL(dl_nns);
 	}
 
       GL(dl_ns)[nsid].libc_map = NULL;
       _dl_debug_change_state (_dl_debug_update (nsid), RT_CONSISTENT);
     }
   /* Never allow loading a DSO in a namespace which is empty.  Such
      direct placements is only causing problems.  Also don't allow
      loading into a namespace used for auditing.  */
   else if (__glibc_unlikely (nsid != LM_ID_BASE && nsid != __LM_ID_CALLER)
 	   && (__glibc_unlikely (nsid < 0 || nsid >= GL(dl_nns))
 	       /* This prevents the [NSID] index expressions from being
 		  evaluated, so the compiler won't think that we are
 		  accessing an invalid index here in the !SHARED case where
 		  DL_NNS is 1 and so any NSID != 0 is invalid.  */
 	       || DL_NNS == 1
 	       || GL(dl_ns)[nsid]._ns_nloaded == 0
 	       || GL(dl_ns)[nsid]._ns_loaded->l_auditing))
     _dl_signal_error (EINVAL, file, NULL,
 		      N_("invalid target namespace in dlmopen()"));
 
   struct dl_open_args args;
   args.file = file;
   args.mode = mode;
   args.nsid = nsid;
   /* args.libc_already_loaded is always assigned by dl_open_worker
      (before any explicit/non-local returns).  */
+
+  args.reported_r_debug_consistent = false;
+  args.relocation_probe_r_debug = NULL;
   args.argc = argc;
   args.argv = argv;
   args.env = env;
 
   /* Determine the caller's map if necessary.  This is needed when we
      don't know the namespace ID in which we have to put the new object,
      in case we have a DST, or when the file name has no path in
      which case we need to look along the RUNPATH/RPATH of the caller.  */
   if (nsid == __LM_ID_CALLER || strchr (file, '$') != NULL
       || strchr (file, '/') == NULL)
     {
       struct dl_find_object dlfo;
       if (_dl_find_object ((void *) caller_dlopen, &dlfo) == 0)
 	args.caller_map = dlfo.dlfo_link_map;
       else
 	/* By default we assume this is the main application.  */
 	args.caller_map = GL(dl_ns)[LM_ID_BASE]._ns_loaded;
       if (args.nsid == __LM_ID_CALLER)
 	args.nsid = args.caller_map->l_ns;
     }
   else
     args.caller_map = NULL;
 
   args.map = _dl_lookup_map (args.nsid, file);
   if (is_already_fully_open (args.map, mode))
     {
       /* We can use the fast path.  */
       ++args.map->l_direct_opencount;
       __rtld_lock_unlock_recursive (GL(dl_load_lock));
       return args.map;
     }
 
   struct dl_exception exception;
   int errcode = _dl_catch_exception (&exception, dl_open_worker, &args);
 
+  maybe_report_r_debug_consistent (&args);
+
   /* Do this for both the error and success cases.  The old value has
      only been determined if the namespace ID was assigned (i.e., it
      is not __LM_ID_CALLER).  In the success case, we actually may
      have consumed more pending adds than planned (because the local
      scopes overlap in case of a recursive dlopen, the inner dlopen
      doing some of the globalization work of the outer dlopen), so the
      old pending adds value is larger than absolutely necessary.
      Since it is just a conservative upper bound, this is harmless.
      The top-level dlopen call will restore the field to zero.  */
   if (args.nsid >= 0)
     GL (dl_ns)[args.nsid]._ns_global_scope_pending_adds
       = args.original_global_scope_pending_adds;
 
   /* See if an error occurred during loading.  */
   if (__glibc_unlikely (exception.errstring != NULL))
     {
       /* Avoid keeping around a dangling reference to the libc.so link
 	 map in case it has been cached in libc_map.  */
       if (!args.libc_already_loaded)
 	GL(dl_ns)[args.nsid].libc_map = NULL;
 
       /* Remove the object from memory.  It may be in an inconsistent
 	 state if relocation failed, for example.  */
       if (args.map)
 	{
 	  _dl_close_worker (args.map, true);
 
 	  /* All l_nodelete_pending objects should have been deleted
 	     at this point, which is why it is not necessary to reset
 	     the flag here.  */
-- 
2.55.0

